File: 1876_countGoodSubstrings.py
class Solution:
    def countGoodSubstrings(self, s: str) -> int:
        distinct = 0
        for i in range(len(s) - 2):
            w = s[i:i+3]
            if w[0] != w[1] and w[0] != w[2]  and w[1] != w[2]:
                distinct += 1
        return distinct

File: 1501-2000/1876_countGoodSubstrings.java
class Solution {
    public int countGoodSubstrings(String s) {
        if (s.length() < 3) return 0;

        HashMap<Character, Integer> window = new HashMap<>();
        int count = 0;

        for (int i=0; i<3; i++) {
            window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);
        }

        if (window.size() == 3) count++;

        for (int i=3; i<s.length(); i++) {
            window.put(s.charAt(i-3), window.get(s.charAt(i-3)) - 1);

            if (window.get(s.charAt(i-3)) == 0) {
                window.remove(s.charAt(i-3));
            }

            window.put(s.charAt(i), window.getOrDefault(s.charAt(i), 0) + 1);
            
            if (window.size() == 3) count++;
        }
        
        return count;
    }
}
